### PR TITLE
Allow passing flexible positions to positional embedding layers

### PR DESCRIPTION
**Pull request by abheesht17**
In DPO, GRPO, etc., we generally want to access only the "completion" logits. In this case, it's helpful to left-pad the prompt, so that the response/completion can be indexed into and accessed. Passing a flexible, batch-wise position tensor is helpful, because every element in the batch will have a different position sequence.

### CODE REVIEW

**Review by divyashreepathihalli — APPROVED** [review]

**Review by mattdangerw — APPROVED** [review]
nice. ty!
